### DISCUSSION TITLE
multi phase grain tracker

### DISCUSSION THREAD

**Discussion by starkekr** [idaholab/moose#22755] [Q&A Modules: Phase field] [answered]
Is there a multi phase grain tracker feature in MOOSE? I am trying to solve a multi-phase (3 phases) multi-grain (hundreds of grains) system and a cant seem to find a multi phase grain tracker object in moose. 

I was thinking of trying to use the GrainDataTracker object to store the phase id but ran into difficulty when initializing the object because the newGrain function only takes as input the grain id. Ideally I would like it to pass the phase as well but i dont know how possible this is seeing as this function derives from newGrainCreated in GrainTracker which i think is pulling data from the FeatureFloodCount object? I guess the alternative is to maintain a grain_id->phase_id map.

Any thoughts on this is greatly appreciated.

**Comment by GiudGiud** [comment, marked as answer]
Hello

Is this page helpful?
https://mooseframework.inl.gov/moose/source/postprocessors/GrainTracker.html
We do support multiphase and multi-grain in the phase field module.

Maybe @permcody could point you further. 

Guillaume

**Reply by starkekr** [reply]
I have read this page. Unfortunately, it was not that useful for incorporating multiphase information into the Grain tracker. The FeatureFloodCount object only seems to have information about the centroid locations and a unique id for the grains. I may be able to use the centroid locations to initialize each of the phases for each grain for my GrainDataTracker object but was wondering if there were any other ideas?

**Reply by permcody** [reply]
Hi @starkekr - As the author of the FeatureFloodCounter and GrainTracker, I did design those objects with was as much flexibility as possible so that developers could easily extend these objects further. The purpose of these two objects is simply to identify "connected regions", and to "track across time" respectively. @dschwen, later created the GrainDataTracker so that we could attach arbitrary information to each grain keyed by unique ID (everything from secondary identifiers, Euler angles, and I'm fairly certain phase should already be included). We should be able to store and retrieve all of that data now. I may have to dig around for an example, but I'm fairly certain there is an EBSD example where we have multiple phases right now. 

Worst case scenario, you can certainly create one more map in your own "GrainTracker-derived" class, but you really shouldn't need to do this. As to your exact question, @dschwen will be able to better answer your question, but rather than pass a bunch of data to the "newGrainCreated" method and inevitably miss something (like phase). I believe we delegated the job of how to create a new grain downstream a bit further so that users could potentially augment the data or method that goes into creating a new grain.

Sorry for all the rambling, I only bring answers to half of the questions here. I do believe we have the capabilities you need, sorry that it's not more clear on how to unlock everything.

**Reply by starkekr** [reply]
This is great information, thank you! The code is written very well! Yes, I see there is an example in ebsd_reconstruction (IN100-111grn.i) which reads grain information from a text file which has the phase id and euler angles as secondary identifiers. But it is still a single phase problem. When I change one of the grains the ebsd reader gives me the following error

"An EBSD feature needs to have a uniform phase."

I dont think we need to dwell on this though because I can write my own object to initialize the grains. Once I hear back from dschwen maybe he can give me a better idea of a path forward.

**Reply by starkekr** [reply]
Just checking to see if there is there any update on this?

**Reply by GiudGiud** [reply]
@dschwen want to comment on what Cody said or on the error @starkekr hit?

**Reply by dschwen** [reply]
If I remember correctly the way this is set up right now is that you have a separate set of OPs for each phase. Each with their own Grain tracker, and each with their own `PolycrystalEBSD` - that user object has a parameter to pick one specific phase from the EBSD file. You'd need to figure out how to formulate the interaction of grains in different phases though.

**Reply by starkekr** [reply]
Is there an example where this is done with the grain tracker? I see there are a few reconstruction tests (some of which use the grain tracker) but these only use one PolycrystalEBSD reader object to initialize the grains (using PolycrystalColoringIC)? 

I have tried creating two sets of variables with different var_base_name's (eta_p1 & eta_p2) then created two PolycrystalEBSD reader objects differing in their phase along with two GrainTracker objects which take in the different PolycrystalEBSD objects. The problem comes in producing the initial conditions for the order parameters.  I have attached the test file which gives me an error that polycrystal_ic_uo and op_num are supplied multiple times. 
[c_chem_only2d_small2.txt](https://github.com/idaholab/moose/files/10385074/c_chem_only2d_small2.txt)

I can write an IC object too but just wanted to check if there was a workaround for this part. 
Also how do I tell the grain tracker to only work on a subset of the order parameters?

**Reply by starkekr** [reply]
Sorry refer to this file. I think I found out how to make the grain tracker work on a subset of order parameters. Is the way I am doing it correct? 

[c_chem_only2d_small2.txt](https://github.com/idaholab/moose/files/10385341/c_chem_only2d_small2.txt)

**Reply by permcody** [reply]
Yes, you can certainly supply the list of variables manually to the GrainTracker. We normally build the list automatically based on a prefix and suffix pattern, but this is valid too.

**Reply by starkekr** [reply]
Is there a way to print out the materials that are defined? I have created an input file which produces the following error:
*** ERROR ***
The following error occurred in the object "EM", of type "ComputeRotatedRankTwoTensorGT".

Failed to get a parameter with the name "grain_trackerMT" when getting a UserObjectName.
[min_error_file.txt](https://github.com/idaholab/moose/files/10491789/min_error_file.txt)

ComputeRotatedRankTwoTensorGT is a templated class which derives from CompositeTensorBase.In this class I try to get one of the created grain trackers using the line

_grain_trackers[i] = &this->template getUserObject<GrainDataTracker<Point>>(_gt_names[i]);

and I get the above error.

**Reply by GiudGiud** [reply]
There's a debug option to show all material properties
see these options
https://mooseframework.inl.gov/source/actions/SetupDebugAction.html

**Reply by starkekr** [reply]
Hmm this does not show userobjects for some reason but I think what I really need is a getUserObjectByName function. Do you know if one of these exists? getUserObject just seems to retrieve the name of the userObject in the parameter list and I have a list of userobjects as inputs to the class I am writing so I need to get the user objects by their name or some other means.

**Reply by GiudGiud** [reply]
Yeah User Objects are a different system

There is such a function?
https://mooseframework.inl.gov/docs/doxygen/moose/classUserObjectInterface.html#a2a4a58c6cf2476901fe790249096170e
It's in the UOInterface, which is inherited by plenty of classes that need to interact with UOs

**Reply by starkekr** [reply]
Thanks!

Is there a way to transfer the grain tracker to a subapp?

**Reply by starkekr** [reply]
I see that there are user object transfers but do I have to provide a way to copy the data?

**Reply by GiudGiud** [reply]
You cant transfer a user object per say. You can transfer using the MultiAppUserObjectTransfer the spatial values that an UO computes over the mesh.

To transfer a grain tracker I imagine you have to transfer all its underlying data item by item

**Reply by permcody** [reply]
@GiudGiud - is correct. The normal way you transfer is moving just data: fields, scalars, other aggregates. So for a Multiapp simulation where you wanted to pass information around you'd normally just pass along data. However, the GrainTracker does create a lot of useful information. In fact, we routinely pull a lot of that data out through the "VectorPostprocessor" system. A similar strategy could be used to pass along all of that aggregated data to the next level down.
